external/sql/storm-sql-core/src/jvm/org/apache/storm/sql/compiler/backends/trident/TridentLogicalPlanCompiler.java

```diff
@@ -0,0 +1,344 @@
+                    .project(new Fields(temporaryOutputFieldNames))
+                    .each(new Fields(temporaryOutputFieldNames), new ForwardFunction(), new Fields(outputFieldNames))
+                    .project(new Fields(outputFieldNames))
+                    .name(stageName);
+        }
+    }
+
+    @Override
+    public IAggregatableStream visitFilter(Filter filter, List<IAggregatableStream> inputStreams) throws Exception {
+        if (inputStreams.size() > 1) {
+            throw new RuntimeException("Filter is a SingleRel");
+        }
+
+        Stream inputStream = inputStreams.get(0).toStream();
+        String stageName = getStageName(filter);
+
+        try (StringWriter sw = new StringWriter(); PrintWriter pw = new PrintWriter(sw)) {
+            pw.write("import org.apache.storm.tuple.Values;\n");
+
+            ExprCompiler compiler = new ExprCompiler(pw, typeFactory);
+            String ret = filter.getCondition().accept(compiler);
+            pw.write(String.format("\nreturn %s;", ret));
+            final String expression = sw.toString();
+
+            return inputStream.filter(new EvaluationFilter(expression)).name(stageName);
+        }
+    }
+
+    @Override
+    public IAggregatableStream visitAggregate(Aggregate aggregate, List<IAggregatableStream> inputStreams) throws Exception {
+        if (inputStreams.size() > 1) {
+            throw new RuntimeException("Aggregate is a SingleRel");
+        }
+
+        Stream inputStream = inputStreams.get(0).toStream();
+        String stageName = getStageName(aggregate);
+
+        List<String> groupByFieldNames = new ArrayList<>();
+        for (Integer idx : aggregate.getGroupSet()) {
+            String fieldName = inputStream.getOutputFields().get(idx);
+            groupByFieldNames.add(fieldName);
+        }
+
+        Fields groupByFields = new Fields(groupByFieldNames);
+        GroupedStream groupedStream = inputStream.groupBy(groupByFields);
+
+        List<Stream> joiningStreams = new ArrayList<>();
+        List<Fields> joiningFields = new ArrayList<>();
+        List<String> newOutputFields = new ArrayList<>();
+        newOutputFields.addAll(groupByFieldNames);
+
+        for (AggregateCall call : aggregate.getAggCallList()) {
+            Stream aggregateStream = handleAggregateCall(groupedStream, groupByFieldNames, inputStream, call);
+
+            Fields outputFields = aggregateStream.getOutputFields();
+            // assuming handleAggregateCall does project so that new output field is always placed to end of fields
+            String newFieldName = outputFields.get(outputFields.size() - 1);
+
+            joiningStreams.add(aggregateStream);
+            joiningFields.add(groupByFields);
+            newOutputFields.add(newFieldName);
+        }
+
+        return topology.join(joiningStreams, joiningFields, new Fields(newOutputFields), JoinType.INNER).name(stageName);
```

Comment:
here we create a separate stream per aggregate function and thats why the join here ? May be we can just execute the different aggregate functions in a single Function and emit the results to avoid the separate streams and join. It may produce optimized topology, but I am not sure.